File: library/src/main/java/com/github/fragivity/Fragivity.java
package com.github.fragivity;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Supplier;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentProviderMap;
import androidx.navigation.fragment.NavHostFragment;


import com.github.fragivity.dialog.DialogUtil;

import java.util.Objects;

import kotlin.jvm.JvmClassMappingKt;
import kotlin.reflect.KClass;

public class Fragivity {

    public static <T extends Fragment> void loadRoot(
            @NonNull NavHostFragment navHost,
            @NonNull Class<T> fragmentClazz) {
        KClass<T> kClass = toKClass(fragmentClazz);
        FragivityUtil.loadRoot(navHost, kClass);
    }

    public static <T extends Fragment> void loadRoot(
            @NonNull NavHostFragment navHost,
            @NonNull Class<T> fragmentClazz,
            @NonNull Supplier<T> factory) {
        KClass<T> kClass = toKClass(fragmentClazz);
        FragmentProviderMap.INSTANCE.put(Objects.requireNonNull(kClass.getQualifiedName()), () -> factory.get());
        FragivityUtil.loadRoot(navHost, kClass);
    }


    public static Navigator of(@NonNull Fragment fragment) {
        return new Navigator(fragment);
    }

    public static class Navigator {

        private MyNavHost myNavHost;

        public Navigator(Fragment fragment) {
            myNavHost = FragivityUtil.getNavigator(fragment);
        }

        public static NavOptionsBuilder navOptionsBuilder() {
            return new NavOptionsBuilder();
        }

        public <T extends Fragment> void push(@NonNull Class<T> fragmentClazz) {
            push(fragmentClazz, null);
        }

        public <T extends Fragment> void push(@NonNull Class<T> fragmentClazz,
                                              @Nullable NavOptions navOptions) {
            FragivityUtil.pushInternal(myNavHost, toKClass(fragmentClazz), navOptions);
        }

        public <T extends Fragment> void push(@NonNull Class<T> fragmentClazz,
                                              @NonNull Supplier<T> factory,
                                              @Nullable NavOptions navOptions) {
            FragmentProviderMap.INSTANCE.put(
                    Objects.requireNonNull(toKClass(fragmentClazz).getQualifiedName()), () -> factory.get());
            push(fragmentClazz, navOptions);
        }

        public <T extends DialogFragment> void showDialog(@NonNull Class<T> fragmentClazz, Bundle args) {
            DialogUtil.showDialog(myNavHost, toKClass(fragmentClazz), args);
        }

    }

    private static <T> KClass<T> toKClass(Class<T> clazz) {
        return JvmClassMappingKt.getKotlinClass(clazz);
    }

}

File: library/src/main/java/com/github/fragivity/Fragivity.kt
package com.github.fragivity

import android.os.Bundle
import androidx.arch.core.util.Function
import androidx.fragment.app.*
import androidx.navigation.fragment.NavHostFragment
import com.github.fragivity.dialog.showDialog

object Fragivity {

    fun setNavOptionsFactory(factory: NavOptions.Factory?) {
        NavOptions.setNavOptionsFactory(factory)
    }

    @JvmStatic
    fun proxyFragmentFactory(activity: FragmentActivity) {
        activity.proxyFragmentFactory()
    }

    @JvmStatic
    fun proxyFragmentFactory(manager: FragmentManager) {
        manager.proxyFragmentFactory()
    }

    @JvmStatic
    fun setupReportFragmentManager(navHost: NavHostFragment) {
        navHost.setupReportFragmentManager()
    }

    @JvmStatic
    fun <T : Fragment> loadRoot(
        navHost: NavHostFragment,
        fragmentClazz: Class<T>
    ) {
        navHost.loadRoot(fragmentClazz.kotlin)
    }

    @JvmStatic
    fun <T : Fragment> loadRoot(
        navHost: NavHostFragment,
        route: String,
        fragmentClazz: Class<T>
    ) {
        navHost.loadRoot(route, fragmentClazz.kotlin)
    }

    @JvmStatic
    fun <T : Fragment> loadRoot(
        navHost: NavHostFragment,
        fragmentClazz: Class<T>,
        factory: Function<Bundle, T>
    ) {
        navHost.loadRoot(fragmentClazz.kotlin) { factory.apply(it) }
    }

    @JvmStatic
    fun <T : Fragment> loadRoot(
        navHost: NavHostFragment,
        route: String,
        fragmentClazz: Class<T>,
        factory: Function<Bundle, T>
    ) {
        navHost.loadRoot(route, fragmentClazz.kotlin) { factory.apply(it) }
    }

    @JvmStatic
    fun composable(
        navHostFragment: NavHostFragment,
        route: String,
        factory: Function<Bundle, Fragment>
    ) {
        navHostFragment.composable(route) { factory.apply(it) }
    }

    @JvmStatic
    fun composable(
        navHostFragment: NavHostFragment,
        route: String,
        argument: NamedNavArgument,
        factory: Function<Bundle, Fragment>
    ) {
        navHostFragment.composable(route, argument) { factory.apply(it) }
    }

    @JvmStatic
    fun composable(
        navHostFragment: NavHostFragment,
        route: String,
        arguments: List<NamedNavArgument>,
        factory: Function<Bundle, Fragment>
    ) {
        navHostFragment.composable(route, arguments) { factory.apply(it) }
    }

    @JvmStatic
    fun of(fragment: Fragment): Navigator {
        return Navigator(fragment)
    }

    class Navigator(val _fragment: Fragment) {

        @JvmOverloads
        fun <T : Fragment> push(
            fragmentClazz: Class<T>,
            navOptions: NavOptions? = null
        ) {
            _fragment.navigator.push(fragmentClazz.kotlin, navOptions)
        }

        @JvmOverloads
        fun <T : Fragment> push(
            fragmentClazz: Class<T>,
            factory: Function<Bundle, T>,
            navOptions: NavOptions? = null
        ) {
            _fragment.navigator.push(fragmentClazz.kotlin, navOptions) { factory.apply(it) }
        }

        @JvmOverloads
        fun <T : Fragment> pushTo(
            fragmentClazz: Class<T>,
            navOptions: NavOptions? = null
        ) {
            _fragment.navigator.pushTo(fragmentClazz.kotlin, navOptions)
        }

        @JvmOverloads
        fun <T : Fragment> pushTo(
            fragmentClazz: Class<T>,
            factory: Function<Bundle, T>,
            navOptions: NavOptions? = null
        ) {
            _fragment.navigator.pushTo(fragmentClazz.kotlin, navOptions) { factory.apply(it) }
        }

        @JvmOverloads
        fun <T : DialogFragment> showDialog(
            fragmentClazz: Class<T>,
            args: Bundle? = null
        ) {
            _fragment.navigator.showDialog(requireNotNull(fragmentClazz.kotlin), args)
        }

        @JvmOverloads
        fun push(route: String, navOptions: NavOptions? = null) {
            _fragment.navigator.push(route, navOptions)
        }

        fun pop(): Boolean {
            return _fragment.navigator.pop()
        }

        @JvmOverloads
        fun popTo(route: String, inclusive: Boolean = false): Boolean {
            return _fragment.navigator.popTo(route, inclusive)
        }

        companion object {
            @JvmStatic
            fun navOptionsBuilder(): NavOptionsBuilder {
                return NavOptionsBuilder()
            }
        }
    }
}